### ISSUE TITLE
ajv cannot load pointer hashes from remote JSON schema

### ISSUE THREAD

**Issue by xp4xbox** [ajv-validator/ajv#1849]
I am using the latest version of ajv.

When I am compiling the following schema:

```
{ title: 'Device Object',
description: 'Defines the Device Model',
type: 'object',
$id: 'https://catena.rossvideo.com/schemata/ogp.schema.json',
properties:
 { slot: { '$ref': '#/$defs/slot' }}
required: [ 'slot' ],
additionalProperties: false }}
```

`AnySchema https://catena.rossvideo.com/schemata/ogp.schema.json is loaded but https://catena.rossvideo.com/schemata/ogp.schema.json#/$defs/slot cannot be resolved`

where `ogp.schema.json` is (from a private domain) and is being loaded under the loadSchema method.

```
{
$schema: "https://json-schema.org/draft-07/schema#",
title: "OGP Schema",
description: "OGP Schema",
type: "boolean",
default: true,
$defs: {
slot: {
title: "Slot ID",
description: "Uniquely identifies the device within the scope of its frame",
type: "integer",
min: 0,
max: 255,
default: 0,
$comment: "Max slot index is set abitrarily, it could go higher if needed."
}
.....
```

**Comment by MatthewPinheiro** [comment]
This is not a bug. From [this docs page](https://ajv.js.org/guide/managing-schemas.html#cache-key-schema-vs-key-vs-id):
> use schema root $id attribute. While it usually looks like URI, it does not mean Ajv downloads it from this URI - this is simply $id used to identify and access the schema. You can though configure Ajv to download schemas on demand - see [Asynchronous schema loading](https://ajv.js.org/guide/managing-schemas.html#asynchronous-schema-loading)

A `$id` can look like just about any string value. It doesn't have to be a traditional URI, but URIs are already a good structure for standardizing the placement of things.

**Comment by xp4xbox** [comment]
edited original

**Comment by MatthewPinheiro** [comment]
If I'm understanding correctly, then you're misusing `$id`. You have two schemas, and you want the first, which is stored locally, to reference the second, which is stored remotely. The problem is, these are two different schemas, but you're trying to use the same `$id` to refer to both of them.

`#` points to either the last ancestor you've assigned an `$id`, or to the root of your schema object. When Ajv compiles your first schema, it reads `{ $ref: '#/$defs/slot' }` and correctly resolves `#/` to the root of your first schema (which you gave an `$id`). So then it looks at the root of your first schema and says, "I don't see a `$defs` property here", and so the resolution fails.

If you instead changed the `$ref` to `https://catena.rossvideo.com/schemata/ogp.schema.json#/$defs/slot`, it will _still_ resolve everything up to the `#` to the root of your first schema, because you've explicitly given it that `$id` already. So again, resolution fails. In fact, your `loadSchema` function is never even called because everything's resolving to the first schema, which you can see in [this runkit](https://runkit.com/matthewpinheiro/ajv-incorrect-id-use/1.0.0).

To get the behavior you expect, change your `$ref` to `https://catena.rossvideo.com/schemata/ogp.schema.json#/$defs/slot` as mentioned above, but also give your first schema a unique `$id` (or just omit it) so there won't be a collision on namespace.

**Comment by xp4xbox** [comment]
I tried doing that but it gives me the following:

`can't resolve reference https://catena.rossvideo.com/schemata/ogp.schema.json#/$defs/slot from id #`

 (i removed the ID)

**Comment by MatthewPinheiro** [comment]
I'm not sure what your data looks like, or what else you are or aren't doing in your code. All I can do is link you [another runkit example](https://runkit.com/matthewpinheiro/ajv-corrected-id-use/1.0.1) showing that this compiles and validates correctly, and that as far as I can tell, this isn't a bug.

**Comment by xp4xbox** [comment]
I understand now, I fixed it thanks.
